File: internal/contacts/googlecsv.go
package contacts

import (
	"encoding/csv"
	"fmt"
	"strings"
)

// ParseGoogleCSV parses a Google Contacts CSV export into Fields records.
//
// Google Contacts exports a file whose first row is a header with columns like:
//
//	Name, Given Name, Family Name, E-mail 1 - Value, Phone 1 - Value,
//	Organization 1 - Name, Organization 1 - Title, Notes, Labels
//
// Multi-value columns follow the pattern "E-mail N - Value" and
// "Phone N - Value" for N = 1, 2, 3 … Google includes up to 3 by default
// but may include more.  Any column that is absent is silently ignored.
//
// The returned slice contains one Fields per data row that carries at least
// one meaningful value; entirely blank rows are skipped.
func ParseGoogleCSV(data string) ([]Fields, error) {
	r := csv.NewReader(strings.NewReader(data))
	r.FieldsPerRecord = -1 // variable column count is OK
	r.LazyQuotes = true    // Google sometimes produces sloppy quoting
	r.TrimLeadingSpace = true

	records, err := r.ReadAll()
	if err != nil {
		return nil, fmt.Errorf("googlecsv: CSV parse error: %w", err)
	}
	if len(records) < 2 {
		// No header or no data rows.
		return nil, nil
	}

	header := records[0]
	colIndex := make(map[string]int, len(header))
	for i, h := range header {
		colIndex[strings.TrimSpace(h)] = i
	}

	// Helper: safely get a cell by column name (empty string if absent/out-of-range).
	get := func(row []string, col string) string {
		idx, ok := colIndex[col]
		if !ok || idx >= len(row) {
			return ""
		}
		return strings.TrimSpace(row[idx])
	}

	var out []Fields
	for _, row := range records[1:] {
		f := Fields{}

		f.DisplayName = get(row, "Name")
		f.FirstName = get(row, "Given Name")
		f.LastName = get(row, "Family Name")
		f.Notes = get(row, "Notes")

		// Some Google exports use "Labels" (single column), others use "Group Membership".
		if lv := get(row, "Labels"); lv != "" {
			for _, l := range splitLabels(lv) {
				f.Labels = appendUniq(f.Labels, l)
			}
		}
		if gm := get(row, "Group Membership"); gm != "" {
			for _, l := range splitGoogleGroups(gm) {
				f.Labels = appendUniq(f.Labels, l)
			}
		}

		// Company / job title from the first organization block.
		f.Company = get(row, "Organization 1 - Name")
		f.JobTitle = get(row, "Organization 1 - Title")

		// Multi-value emails: "E-mail 1 - Value", "E-mail 2 - Value", …
		for n := 1; n <= 10; n++ {
			v := get(row, fmt.Sprintf("E-mail %d - Value", n))
			if v == "" {
				break
			}
			f.Emails = appendUniq(f.Emails, v)
		}

		// Multi-value phones: "Phone 1 - Value", …
		for n := 1; n <= 10; n++ {
			v := get(row, fmt.Sprintf("Phone %d - Value", n))
			if v == "" {
				break
			}
			f.Phones = appendUniq(f.Phones, v)
		}

		// Derive display name if absent.
		if f.DisplayName == "" {
			full := strings.TrimSpace(f.FirstName + " " + f.LastName)
			if full != "" {
				f.DisplayName = full
			} else if len(f.Emails) > 0 {
				f.DisplayName = f.Emails[0]
			}
		}

		if isMeaningfulFields(f) {
			out = append(out, f)
		}
	}
	return out, nil
}

// isMeaningfulFields returns true when the record carries enough information
// to be worth creating as a contact.
func isMeaningfulFields(f Fields) bool {
	return f.DisplayName != "" || f.FirstName != "" || f.LastName != "" ||
		len(f.Emails) > 0 || len(f.Phones) > 0
}

// appendUniq appends v to s only when v is non-empty and not already present
// (case-insensitive comparison).
func appendUniq(s []string, v string) []string {
	v = strings.TrimSpace(v)
	if v == "" {
		return s
	}
	vl := strings.ToLower(v)
	for _, x := range s {
		if strings.ToLower(x) == vl {
			return s
		}
	}
	return append(s, v)
}

// splitLabels handles a Google "Labels" cell which uses " ::: " as a separator
// in some exports (e.g. "myContacts ::: Work") or simple commas in others.
func splitLabels(v string) []string {
	var parts []string
	// Try the triple-colon separator first (Google's Group Membership format).
	if strings.Contains(v, ":::") {
		for _, p := range strings.Split(v, ":::") {
			p = strings.TrimSpace(p)
			if p != "" {
				parts = append(parts, p)
			}
		}
		return parts
	}
	// Fall back to comma-separated.
	for _, p := range strings.Split(v, ",") {
		p = strings.TrimSpace(p)
		if p != "" {
			parts = append(parts, p)
		}
	}
	return parts
}

// splitGoogleGroups splits a "Group Membership" cell which Google separates
// with " ::: " and may include the built-in group "* myContacts" (prefixed
// with an asterisk).  We strip the asterisk prefix and skip the "myContacts"
// default group since it maps to no user-visible label.
func splitGoogleGroups(v string) []string {
	var parts []string
	for _, p := range strings.Split(v, ":::") {
		p = strings.TrimSpace(p)
		p = strings.TrimPrefix(p, "* ") // strip Google's asterisk marker
		if p == "" || strings.EqualFold(p, "myContacts") || strings.EqualFold(p, "my contacts") {
			continue
		}
		parts = append(parts, p)
	}
	return parts
}

File: web/app/src/pages/contacts/googlecsv.ts
import type { ParsedContact } from "./vcard";

/**
 * Parse a Google Contacts CSV export into ParsedContact records.
 *
 * Google Contacts exports a CSV whose first row is a header.  The columns of
 * interest are:
 *
 *   Name, Given Name, Family Name, Notes, Group Membership,
 *   E-mail N - Value  (N = 1, 2, 3 …),
 *   Phone N - Value   (N = 1, 2, 3 …),
 *   Organization 1 - Name, Organization 1 - Title
 *
 * Multi-value columns stop as soon as an empty value is encountered.
 * The "Group Membership" column uses " ::: " as a separator and may contain
 * a built-in "* myContacts" entry that is silently dropped.
 */
export function parseGoogleCSV(text: string): ParsedContact[] {
  const rows = parseCSVRows(text);
  if (rows.length < 2) return [];

  const header = rows[0].map((h) => h.trim());

  const col = (name: string): number => header.indexOf(name);

  // Build a getter: returns cell value (trimmed) or "".
  const get = (row: string[], name: string): string => {
    const idx = col(name);
    if (idx < 0 || idx >= row.length) return "";
    return row[idx].trim();
  };

  const out: ParsedContact[] = [];

  for (const row of rows.slice(1)) {
    const displayName = get(row, "Name");
    const firstName = get(row, "Given Name");
    const lastName = get(row, "Family Name");
    const notes = get(row, "Notes");
    const company = get(row, "Organization 1 - Name");
    const jobTitle = get(row, "Organization 1 - Title");

    const emails: string[] = [];
    const phones: string[] = [];

    // Collect multi-value emails: stop at first empty slot.
    for (let n = 1; n <= 20; n++) {
      const v = get(row, `E-mail ${n} - Value`);
      if (!v) break;
      uniqPush(emails, v);
    }

    // Collect multi-value phones.
    for (let n = 1; n <= 20; n++) {
      const v = get(row, `Phone ${n} - Value`);
      if (!v) break;
      uniqPush(phones, v);
    }

    // Labels from "Group Membership" (Google) or "Labels" (simplified export).
    const labels: string[] = [];
    const gm = get(row, "Group Membership") || get(row, "Labels");
    if (gm) {
      const parts = gm.includes(":::") ? gm.split(":::") : gm.split(",");
      for (const p of parts) {
        // Google prefixes built-in groups with "* " — strip it and skip myContacts.
        const clean = p.trim().replace(/^\* /, "").trim();
        if (!clean || /^my\s*contacts$/i.test(clean)) continue;
        uniqPush(labels, clean);
      }
    }

    // Derive display name if absent.
    let derived = displayName;
    if (!derived) {
      const full = `${firstName} ${lastName}`.trim();
      derived = full || emails[0] || "";
    }

    if (!isMeaningful(derived, firstName, lastName, emails, phones)) continue;

    out.push({
      display_name: derived,
      first_name: firstName,
      last_name: lastName,
      company,
      job_title: jobTitle,
      emails,
      phones,
      labels,
      notes,
    });
  }

  return out;
}

function isMeaningful(
  displayName: string,
  firstName: string,
  lastName: string,
  emails: string[],
  phones: string[],
): boolean {
  return Boolean(
    displayName || firstName || lastName || emails.length || phones.length,
  );
}

function uniqPush(arr: string[], v: string): void {
  const t = v.trim();
  if (t && !arr.some((x) => x.toLowerCase() === t.toLowerCase())) arr.push(t);
}

// ---------------------------------------------------------------------------
// Minimal RFC 4180 CSV parser (handles quoted fields, embedded commas/newlines,
// doubled-quote escaping, CRLF and LF line endings).
// ---------------------------------------------------------------------------
function parseCSVRows(text: string): string[][] {
  const rows: string[][] = [];
  let cur: string[] = [];
  let field = "";
  let inQuote = false;
  const n = text.length;

  for (let i = 0; i < n; i++) {
    const ch = text[i];

    if (inQuote) {
      if (ch === '"') {
        // Peek ahead: "" is an escaped quote inside a quoted field.
        if (i + 1 < n && text[i + 1] === '"') {
          field += '"';
          i++; // consume second quote
        } else {
          inQuote = false; // closing quote
        }
      } else {
        field += ch;
      }
      continue;
    }

    // Not in a quoted field.
    if (ch === '"') {
      inQuote = true;
      continue;
    }
    if (ch === ",") {
      cur.push(field);
      field = "";
      continue;
    }
    if (ch === "\r") {
      // Handle CRLF: peek and consume the \n.
      if (i + 1 < n && text[i + 1] === "\n") i++;
      cur.push(field);
      field = "";
      rows.push(cur);
      cur = [];
      continue;
    }
    if (ch === "\n") {
      cur.push(field);
      field = "";
      rows.push(cur);
      cur = [];
      continue;
    }
    field += ch;
  }

  // Flush any remaining field/row.
  cur.push(field);
  if (cur.some((f) => f !== "")) rows.push(cur);

  return rows;
}
